herddb-core/src/main/java/herddb/cluster/BookKeeperDataStorageManager.java

```diff
@@ -0,0 +1,1744 @@
+        }
+        dataPageWrites.registerSuccessfulEvent(delta, TimeUnit.MILLISECONDS);
+    }
+
+    private static long writeIndexPage(DataWriter writer, VisibleByteArrayOutputStream stream) throws IOException {
+        try (
+                ExtendedDataOutputStream dataOutput = new ExtendedDataOutputStream(stream)) {
+            dataOutput.writeVLong(1); // version
+            dataOutput.writeVLong(0); // flags for future implementations
+            writer.write(dataOutput);
+            dataOutput.flush();
+            long hash = XXHash64Utils.hash(stream.getBuffer(), 0, stream.size());
+            dataOutput.writeLong(hash);
+            dataOutput.flush();
+            return stream.size();
+        }
+    }
+
+    @Override
+    public void writeIndexPage(
+            String tableSpace, String indexName,
+            long pageId, DataWriter writer
+    ) throws DataStorageManagerException {
+        long _start = System.currentTimeMillis();
+
+        long size;
+        long ledgerId;
+        try {
+            try (VisibleByteArrayOutputStream buffer = new VisibleByteArrayOutputStream()) {
+
+                size = writeIndexPage(writer, buffer);
+                Map<String, byte[]> metadata = new HashMap<>();
+                metadata.put("tablespaceuuid", tableSpace.getBytes(StandardCharsets.UTF_8));
+                metadata.put("node", nodeId.getBytes(StandardCharsets.UTF_8));
+                metadata.put("application", "herddb".getBytes(StandardCharsets.UTF_8));
+                metadata.put("component", "datastore".getBytes(StandardCharsets.UTF_8));
+                metadata.put("index", indexName.getBytes(StandardCharsets.UTF_8));
+                metadata.put("type", "indexpage".getBytes(StandardCharsets.UTF_8));
+                try (WriteHandle result =
+                        FutureUtils.result(bk.getBookKeeper()
+                                .newCreateLedgerOp()
+                                .withEnsembleSize(bk.getEnsemble())
+                                .withWriteQuorumSize(bk.getWriteQuorumSize())
+                                .withAckQuorumSize(bk.getAckQuorumSize())
+                                .withPassword(new byte[0])
+                                .withDigestType(DigestType.CRC32C)
+                                .withCustomMetadata(metadata)
+                                .execute(), BKException.HANDLER);) {
+                    result.append(buffer.getBuffer(), 0, buffer.size());
+                    ledgerId = result.getId();
+                }
+            }
+
+        } catch (IOException | org.apache.bookkeeper.client.api.BKException err) {
+            throw new DataStorageManagerException(err);
+        } catch (InterruptedException err) {
+            Thread.currentThread().interrupt();
+            throw new DataStorageManagerException(err);
+        }
+        getTableSpacePagesMapping(tableSpace).getIndexPagesMapping(indexName).writePageId(pageId, ledgerId);
```

Comment:
BLink trees keep a two page id sets: one is the flush set and can be overridden the other is the checkpoint set and is never overridden. Both use writeIndexPage then here we could overwrite page to ledger knowledge and forget the old ledger (and miss his deletion on a future checkpoint when the ledger itself won't be needed anymore)